### ISSUE TITLE
Dependency injection without default constructor

### ISSUE THREAD

**Issue by natros** [gilberto-torrezan/gwt-views#3]
In this view I don't have a default constructor but I'm using constructor injection.

``` java
@View(value = "login", publicAccess = true, defaultView = true, usesViewContainer = false,
        injector = BackOfficeInjector.class, injectorMethod = "getLoginView")
public class LoginView extends Composite {
    @Inject
    public LoginView(Alert alert) {
        Binder ui = GWT.create(Binder.class);
        initWidget(ui.createAndBindUi(this));
    }
```

```
Compiling module pt.ipb.candidaturas.bo.BackOffice
[INFO]          Computing all possible rebind results for 'com.github.gilbertotorrezan.gwtviews.client.INavigationManager'
[INFO]             Rebinding com.github.gilbertotorrezan.gwtviews.client.INavigationManager
[INFO]                Invoking generator com.github.gilbertotorrezan.gwtviews.rebind.NavigationManagerGenerator
[INFO]                   [WARN] LoginView must have an empty constructor to be a valid View.
[INFO]                   [ERROR] No default view page defined!
[INFO]          [ERROR] Errors in 'com/github/gilbertotorrezan/gwtviews/client/NavigationManager.java'
[INFO]             [ERROR] Line 40: Failed to resolve 'com.github.gilbertotorrezan.gwtviews.client.INavigationManager' via deferred binding
[INFO]          [WARN] For the following type(s), generated source was never committed (did you forget to call commit()?)
[INFO]             [WARN] com.github.gilbertotorrezan.gwtviews.client.INavigationManagerImpl
[INFO]          Unification traversed 11291 fields and methods and 957 types. 932 are considered part of the current module and 932 had all of their fields and methods traversed.
```

**Comment by gilberto-torrezan** [comment]
Thanks for pointing that out. I changed the way the empty constructor requirement is evaluated, and now it allows any type of constructors on Views and ViewContainers when there's an injector or a custom Presenter declared at the annotation. The fix is on the v.1.1.1.

**Comment by natros** [comment]
Hi again.

I'm running super dev mode and when I add a injector for a new view the compiler complains about a missing getter for the view.

``` java
@View(value = "users", rolesAllowed = "USERS", injector = BackOfficeInjector.class, injectorMethod = "getUsersView")
public class UsersView extends Composite {
    @Inject
    public UsersView(Binder ui) {
        initWidget(ui.createAndBindUi(this));
    }
```

``` java
@GinModules(BackOfficeModule.class)
public interface BackOfficeInjector extends Ginjector {
    BackOfficeView getBackOfficeView();

    LoginView getLoginView();

    HomeView getHomeView();

    UsersView getUsersView();

    void inject(BackOfficeApp app);
}
```

```
[INFO] GET /recompile/bo
[INFO]    Job pt.ipb.candidaturas.bo.BackOffice_2_13
[INFO]       starting job: pt.ipb.candidaturas.bo.BackOffice_2_13
[INFO]       Compiling module pt.ipb.candidaturas.bo.BackOffice
[INFO]          Computing all possible rebind results for 'com.github.gilbertotorrezan.gwtviews.client.INavigationManagerImpl.UsersViewPresenter'
[INFO]             Rebinding com.github.gilbertotorrezan.gwtviews.client.INavigationManagerImpl.UsersViewPresenter
[INFO]                Invoking generator com.github.gilbertotorrezan.gwtviews.rebind.PresenterGenerator
[INFO]                   The injector pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector has no methods with pt.ipb.candidaturas.bo.client.views.users.UsersView as return type. The View will not be injected.
[INFO]          [ERROR] Errors in 'gen/pt/ipb/candidaturas/bo/client/views/users/UsersView_UsersViewPresenterImpl.java'
[INFO]             [ERROR] Line 20: Rebind result 'pt.ipb.candidaturas.bo.client.views.users.UsersView' has no default (zero argument) constructors
[INFO]          Unification traversed 671 fields and methods and 320 types. 21 are considered part of the current module and 21 had all of their fields and methods traversed.
[INFO]       [ERROR] Compiler returned false
[INFO]       [WARN] recompile failed
[INFO]       [WARN] continuing to serve previous version
```

I have to restart the super dev mode to overcome the issue.

For some reason the generator is not picking up the getUsersView from injector.

**Comment by gilberto-torrezan** [comment]
Do you have to restart the super dev mode at every modification, or that happened only once? It seems to me that the project was using the old v.1.1.0 somehow, and restarting it made it work with the right version.

If you are using Maven, I recommend you to refresh the Maven dependencies on your IDE after changing versions, and restarting the super dev mode as well. At the command line, one thing you can do to make sure you're using the right version of the lib is to built the project using:

``` bash
mvn clean package -U
```

**Comment by natros** [comment]
That's not the case. Right now I have dependency injection working with the views, but as soon as I add a new view with injector and injectorMethod it complains about a missing getter for the view in the injector. But the method is there and works. I can call the method in the entry point (injector.getMyNewView) that works. At first I thought it was problem with gin, but injector.getMyNewView() is working.

As I said, only restarting super dev mode helps (no need for mvn clean or install), but I only have to restart for each new view I add.

I'm using gwt maven plugin (https://github.com/tbroyer/gwt-maven-plugin)

**Comment by gilberto-torrezan** [comment]
That's weird... Somehow the incremental compiler isn't detecting changes to the injector, and thus providing the old version to the code generator (that produces the automatic Presenter). I'll make some research to see what can be done.

**Comment by gilberto-torrezan** [comment]
By the way, which IDE, GIN version and GWT version are you using?

**Comment by natros** [comment]
Intellij 14.1.3 ultimate,
gwt 2.7.0
gin 2.1.2

**Comment by natros** [comment]
I created a new view (NotFoundView) and added getNotFoundView() to my injector

``` java
@GinModules(BackOfficeModule.class)
public interface BackOfficeInjector extends Ginjector {
    NotFoundView getNotFoundView();

    BackOfficeView getBackOfficeView();

    LoginView getLoginView();

    HomeView getHomeView();

    UsersView getUsersView();

    void inject(BackOfficeApp app);
}
```

After debugging PresenterGenerator the injector.getDeclaredMethods() does not return getNotFoundView()

```
0 = {Method@6036} "public abstract void pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector.inject(pt.ipb.candidaturas.bo.client.BackOfficeApp)"
1 = {Method@6037} "public abstract pt.ipb.candidaturas.bo.client.views.BackOfficeView pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector.getBackOfficeView()"
2 = {Method@6038} "public abstract pt.ipb.candidaturas.bo.client.views.login.LoginView pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector.getLoginView()"
3 = {Method@6039} "public abstract pt.ipb.candidaturas.bo.client.views.home.HomeView pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector.getHomeView()"
4 = {Method@6040} "public abstract pt.ipb.candidaturas.bo.client.views.users.UsersView pt.ipb.candidaturas.bo.client.gin.BackOfficeInjector.getUsersView()"
```

**Comment by gilberto-torrezan** [comment]
Thanks for the debug info. I have a bad feeling about this: maybe it is a bug of the incremental compiler or the code generator (or IDE plugin). I posted a message to the official mailing list to see if I get any insights: https://groups.google.com/forum/#!topic/google-web-toolkit/yR81CQM1NlI

**Comment by natros** [comment]
I may not understand how a generator works in GWT, but instead of using reflection to find the right method, can't you use the method name that is provided in the annotation?

I have a feeling that the generator must be gin-aware.

**Comment by gilberto-torrezan** [comment]
I can do that, but that check was to ensure that the provided method actually exists, and give a proper error message when it doesn't. But given the nature of the error, I think the problem will propagate to the compilation stage (the class definition won't be updated, and the method will not be found anyway). I have to test.

**Comment by gilberto-torrezan** [comment]
Ok, I think I solved the issue. The problem is you cannot use the standard java.lang.Class and java.land.reflect.\* classes to use reflection when generating code in GWT. You have to use the [TypeOracle](http://www.gwtproject.org/javadoc/latest/com/google/gwt/core/ext/typeinfo/TypeOracle.html), [JClassType](http://www.gwtproject.org/javadoc/latest/com/google/gwt/core/ext/typeinfo/JClassType.html) and derivatives. By using them, the injector class is always up-to-date, without the need of reloading the entire code server.

The fix is released on v.1.1.2.
